### DISCUSSION TITLE
Создание коллекции лейбла

### DISCUSSION THREAD

**Discussion by aazbr** [Chimildic/goofy#152] [Личный опыт]
Добрый день. 
Хотел бы поделиться алгоритмом создания коллекции лейбла и сортировки плейлиста по дате релиза на лейбле.

пример по расширенной форме поиска типа _label:"название лейбла"_

Сам алгоритм:
```
function mcryo() {

let cryo = Source.mineTracks({
  type: ['track'],
  keyword: ['label:"cryo chamber"'],
  requestCount: 20,
  itemCount: 2000,
});

 //Filter.dedupTracks(cryo);
 //Order.sort(cryo, 'album.release_date', 'asc');

    Playlist.saveWithAppend({
        id: 'чччч', // после первого создания плейлиста
        name: 'Cryo Chamber collection',
        tracks: cryo,
        //description: Playlist.getDescription(cryo),
        //randomCover: 'update',
    });
 let cryo2 = Source.getPlaylistTracks('Cryo Chamber collection', '3fjq0nvMB2443iHcBCSHCd');
 Order.sort(cryo2, 'album.release_date', 'desc');
 Filter.dedupTracks(cryo2);

 Playlist.saveWithReplace({
        id: 'яяяя', // после первого создания плейлиста
        name: 'Cryo Chamber collection',
        tracks: cryo2,
        //description: Playlist.getDescription(cryo),
        //randomCover: 'update',
});
}
```

Теперь критика:
Итак, первое что меня смущает - количество запросов, есть ли возможность их как-то сократить ? насколько я понимаю, единственная функция, которая использует расширенный поиск - это "_mineTracks_". Сортировка по дате релиза "'_album.release_date'_" тоже делает дополнительные запросы.
Второе - количество треков результата поиска, насколько я вижу - это 1000 штук.
Я вручную раньше собирал по списку альбомов из поиска типа "label:"cryo chamber"" - у меня такой плейлист занимает 1497 треков, то есть, какие-то потеряны. Именно поэтому я сделал 2 объединения плейлистов, но это ситуацию не решило.
Третий момент - некоторые альбомы перемешаны, т.к. даты релизов одинаковые и, видимо, они корректно не вставляются.

И вопрос, возможно ли выполнять расширенный поиск, но с выдачей альбомов? чтобы можно было собирать прямо альбомы в массив, мне кажется это бы сократило запросы.

**Comment by Chimildic** [comment]
> количество запросов, есть ли возможность их как-то сократить?

Не уточнили о каком количестве запросов речь. С кодом ниже вышло 222 запроса и  1505 треков. Учитывая задачу _собрать все_, нормальные показатели. 

Можно сократить, если уточнить поиск. Например, год или период `label:"cryo chamber" AND year:2021`. По примеру [расширенного поиска](https://support.spotify.com/ru-ru/article/search/).

>  количество треков результата поиска, насколько я вижу - это 1000 штук

Одна тысяча это [лимит на элементы результата от Spotify](https://developer.spotify.com/documentation/web-api/reference/#/operations/search). Но элементы имеют разный тип. Тысяча треков и тысяча альбомов совсем разные вещи. И даже когда у лейбла будет больше тысячи альбомов, их можно собрать все через периоды в запросе поиска.

В вашем коде ошибка при указании типа `['track']`. Здесь не нужен массив, только строка `'track'`.

> некоторые альбомы перемешаны

Насколько я вижу по коду ниже, проблема решается сама собой, когда поиск по типу альбома.

>  выполнять расширенный поиск, но с выдачей альбомов

Все возможности функции `mineTracks` описаны в [документации](https://chimildic.github.io/goofy/#/func?id=minetracks). Включая поиск по альбомам. Сразу же будут запросы на получение треков из альбомов. Поскольку в выдаче поиска о них данных нет. На каждый альбом минимум по одному запросу придется.

Единственное, надо сделать правку в документации, что расширенный поиск работает не только трекам, а и по альбомам тоже.

```js
function updatePlaylistOfLabel() {
  let tracks = Source.mineTracks({
    type: 'album',
    keyword: ['label:"cryo chamber"'],
    requestCount: 20,
    itemCount: 1000,
    inRow: true,
  });

  Order.sort(tracks, 'album.release_date', 'desc');

  Playlist.saveWithReplace({
    name: 'Label Cryo Chamber',
    tracks: tracks,
  })

  console.log(CustomUrlFetchApp.getCountRequest());
}
```

**Reply by aazbr** [reply]
О, спасибо, буду дорабатывать!
С типом альбома теперь должно все работать.

По запросам, я полагал, что 1000 треков - это 1к запросов, плюс запрос даты релиза.
надо было тоже прописать лог, стало бы понятнее.

Думаю, после нг сокращу по текущему году, чтобы снизить еще запросы.
